release-notes/3.0/3.0.0-preview3.md

```diff
@@ -0,0 +1,176 @@
+}
+```
+
+The static API on `SqlConnection`, i.e. `SqlConnection.RegisterColumnEncryptionKeyStoreProviders` to register custom master key store providers globally continues to be supported. The column encryption key cache maintained globally only applies to globally registered providers.
+
+#### Column master key store provider registration precedence
+
+The built-in column master key store providers that are available for the Windows Certificate Store, CNG Store and CSP are pre-registered. No providers should be registered on the connection or command instances if one of the built-in column master key store provider is needed.
+
+Custom master key store providers can be registered with the driver at three different layers. The global level is as it currently is. The new per-connection and per-command level registrations will be empty initially and can be set more than once.
+
+The precedence of the three registrations are as follows:
+
+- The per-command registration will be checked if it is not empty.
+- If the per-command registration is empty, the per-connection registration will be checked if it is not empty.
+- If the per-connection registration is empty, the global registration will be checked.
+
+Once any key store provider is found at a registration level, the driver will **NOT** fall back to the other registrations to search for a provider. If providers are registered but the proper provider is not found at a level, an exception will be thrown containing only the registered providers in the registration that was checked.
+
+#### Column encryption key cache precedence
+
+The column encryption keys (CEKs) for custom key store providers registered using the new instance-level APIs will not be cached by the driver. The key store providers need to implement their own cache to gain performance. This local cache of column encryption keys implemented by custom key store providers will be disabled by the driver if the key store provider instance is registered in the driver at the global level.
+
+A new API has also been introduced on the `SqlColumnEncryptionKeyStoreProvider` base class to set the cache time to live:
+
+```cs
+public abstract class SqlColumnEncryptionKeyStoreProvider
+{
+    // The default value of Column Encryption Key Cache Time to Live is 0.
+    // Provider's local cache is disabled for globally registered providers.
+    // Custom key store provider implementation must include column encryption key cache to provide caching support to locally registered providers.
+    public virtual TimeSpan? ColumnEncryptionKeyCacheTtl { get; set; } = new TimeSpan(0);
+}
+```
+
+### IP Address preference
+
+A new connection property `IPAddressPreference` is introduced to specify the IP address family preference to the driver when establishing TCP connections. Below are the three accepted values for this property:
```

Comment:
Could you add a note about its conflict with `Multi Subnet Failover` and `Transparent Network IP Resolution` settings in .NET Framework and .NET Core.

Reply:
```suggestion
A new connection property `IPAddressPreference` is introduced to specify the IP address family preference to the driver when establishing TCP connections. If `Transparent Network IP Resolution` (in .NET Framework) or `Multi Subnet Failover` is set to `true`, this setting has no effect. Below are the three accepted values for this property:
```